File: decorator/java/DecoratorClient.java
class Component {
  void operation() {
    System.out.println("Component instance has been created");
  }
}

class ConcreteComponent extends Component {
  @Override
  public void operation() {
    System.out.println("DecoratorClient instance has been created");
  }
}

class Decorator extends Component {
  Component component;

  public Decorator(Component component) {
    this.component = component;
  }

  @Override
  public void operation() {
    component.operation();
  }
}

class ConcreteDecoratorA extends Decorator {

  public ConcreteDecoratorA(Component component) {
    super(component);
  }

  void addedState() {
    System.out.println("Added state");
  }
}

class ConcreteDecoratorB extends Decorator {

  public ConcreteDecoratorB(Component component) {
    super(component);
  }

  void addedBehavior() {
    System.out.println("Added behavior");
  }
}

public class DecoratorClient {
  public static void main(String[] args) {
    ConcreteComponent component = new ConcreteComponent();
    ConcreteDecoratorA concreteDecoratorA = new ConcreteDecoratorA(component);
    ConcreteDecoratorB concreteDecoratorB = new ConcreteDecoratorB(component);
    component.operation();
    concreteDecoratorA.addedState();
    concreteDecoratorB.addedBehavior();
  }
}

File: decorator/python/DecoratorClient.py
class Component(object):
    def operation(self):
        print("Component instance has been created")


class ConcreteComponent(Component):
    def operation(self):
        print("ConcreteComponent instance has been created")


class Decorator(Component):
    def __init__(self, component):
        self._component = component

    def operation(self):
        print("Decorator instance has been created")


class ConcreteDecoratorA(Decorator):
    def addedState(self):
        print("Added state")


class ConcreteDecoratorB(Decorator):
    def addedBehavior(self):
        print("Added behavior")


if __name__ == "__main__":
    component = ConcreteComponent()
    componenta = ConcreteDecoratorA(component)
    componentb = ConcreteDecoratorB(component)
    component.operation()
    componenta.addedState()
    componentb.addedBehavior()
